nifi-nar-bundles/nifi-enrich-bundle/nifi-enrich-processors/src/main/java/org/apache/nifi/processors/GeoEnrichIP.java

```diff
@@ -97,10 +112,36 @@ public void onTrigger(final ProcessContext context, final ProcessSession session
             return;
         }
 
-        final StopWatch stopWatch = new StopWatch(true);
+        StopWatch stopWatch = new StopWatch(true);
         try {
             response = dbReader.city(inetAddress);
             stopWatch.stop();
+        } catch (InvalidDatabaseException idbe) {
+            stopWatch.stop();
+            if (dbWriteLock.tryLock()) {
+                try {
+                    getLogger().debug("Attempting to reload database after InvalidDatabaseException");
+                    try {
+                        loadDatabaseFile();
+                    } catch (IOException ioe) {
+                        throw new ProcessException("Error reloading database due to: " + ioe.getMessage(), ioe);
+                    }
+
+                    getLogger().debug("Attempting to retry lookup after InvalidDatabaseException");
+                    try {
+                        dbReader = databaseReaderRef.get();
+                        stopWatch = new StopWatch(true);
+                        response = dbReader.city(inetAddress);
+                        stopWatch.stop();
+                    } catch (final Exception e) {
+                        throw new ProcessException("Error performing look up: " + e.getMessage(), e);
+                    }
+                } finally {
+                    dbWriteLock.unlock();
+                }
+            } else {
+                throw new ProcessException("Failed to lookup the key " + inetAddress + " due to " + idbe.getMessage(), idbe);
+            }
```

Comment:
This is a lot of complexity for Exception handling, and its quite repetitive of what's been done above. I think I'd be tempted to instead re-queue the FlowFIle (either by calling `session.transfer(flowFile);` or `session.rollback();`) and just setting a `needsReload` member variable or something, so that on the next invocation the database gets reloaded. Any thoughts on that approach?